packages/entities/entities-plugins/src/components/free-form/Datakit/schema/shared.ts

```diff
@@ -0,0 +1,128 @@
+import type { z } from 'zod'
+
+export type ConnectionRef = {
+  nodeName: string
+  path: Array<string | number>
+  value: string
+}
+
+export function collectConnectionRefs(
+  node: any,
+  nodeIndex: number,
+): ConnectionRef[] {
+  const out: ConnectionRef[] = []
+  const push = (path: Array<string | number>, v: unknown) => {
+    if (typeof v === 'string') {
+      out.push({
+        nodeName: node.name,
+        path: ['nodes', nodeIndex, ...path],
+        value: v,
+      })
+    }
+  }
+  if ('input' in node && node.input != null) push(['input'], node.input)
+  if ('inputs' in node && node.inputs && typeof node.inputs === 'object') {
+    for (const [k, v] of Object.entries(
+      node.inputs as Record<string, unknown>,
+    )) {
+      if (v != null) push(['inputs', k], v)
+    }
+  }
+  if ('output' in node && node.output != null) push(['output'], node.output)
+  if ('outputs' in node && node.outputs && typeof node.outputs === 'object') {
+    for (const [k, v] of Object.entries(
+      node.outputs as Record<string, unknown>,
+    )) {
+      if (v != null) push(['outputs', k], v)
+    }
+  }
+  return out
+}
+
+export function isOutRef(ref: ConnectionRef) {
+  const slot = ref.path[2] as string
+  return slot === 'output' || slot === 'outputs'
+}
+
+export function hasAnyNonNullEntry(obj: unknown): boolean {
+  if (!obj || typeof obj !== 'object') return false
+  for (const v of Object.values(obj as Record<string, unknown>)) {
+    if (v != null) return true
+  }
+  return false
+}
+
+export function validateNamesAndConnections(
+  config: { nodes?: any[] | null },
```

Comment:
Is the `any[]` type intentional?

Reply:
No and I've removed all unnecessary `any`s and type assertions.